File: examples/LMemLoopback/client/rb/AdvancedStatic/LMemLoopbackClient.rb
#!/usr/bin/env ruby

$LOAD_PATH.push('../gen-rb')

require 'thrift'
require 'l_mem_loopback_service'

def check(data_out, expected, size)
  status = 0
  (0..(size - 1)).each do |i|
    if data_out[i] != expected[i]
      puts "#{data_out[i]} != #{expected[i]}"
      status += 1
    end
  end
  status
end

def lmem_loopback_cpu(in_a, in_b, size, data_out)
  (0..(size - 1)).each do |i|
    data_out[i] = (in_a[i] + in_b[i])
  end
end

begin

  include Com::Maxeler::LMemLoopback

  start_time = Time.now
  start_dfe_time = start_time

  port = 9090

  # Make socket
  socket = Thrift::Socket.new('localhost', port)

  # Buffering is critical. Raw sockets are very slow
  transport = Thrift::BufferedTransport.new(socket)

  # Wrap in a protocol
  protocol = Thrift::BinaryProtocol.new(transport)

  # Create a client to use the protocol encoder
  client = LMemLoopbackService::Client.new(protocol)
  current_time = (Time.now - start_time).round(5)
  puts "Creating a client:\t\t\t\t#{current_time}s"

  # Connect!
  start_time = Time.now
  transport.open
  current_time = (Time.now - start_time).round(5)
  puts "Opening connection:\t\t\t\t#{current_time}s"

  size = 384
  size_bytes = size * 4

  # Generate input
  start_time = Time.now
  in_a = Array.new(size)
  in_b = Array.new(size)

  (0..(size - 1)).each do |i|
    in_a[i] = i
    in_b[i] = size - i
  end
  current_time = (Time.now - start_time).round(5)
  puts "Generating input data:\t\t\t\t#{current_time}s"

  # Initialize maxfile
  start_time = Time.now
  max_file = client.LMemLoopback_init
  current_time = (Time.now - start_time).round(5)
  puts "Initializing maxfile:\t\t\t\t#{current_time}s"

  # Load DFE
  start_time = Time.now
  max_engine = client.max_load(max_file, '*')
  current_time = (Time.now - start_time).round(5)
  puts "Loading DFE:\t\t\t\t\t#{current_time}s"

  # Allocate and send input streams to server
  start_time = Time.now
  address_in_a = client.malloc_int32_t(size)
  client.send_data_int32_t(address_in_a, in_a)

  address_in_b = client.malloc_int32_t(size)
  client.send_data_int32_t(address_in_b, in_b)
  current_time = (Time.now - start_time).round(5)
  puts "Sending input data:\t\t\t\t#{current_time}s"

  # Allocate memory for output stream on server
  start_time = Time.now
  address_out_data = client.malloc_int32_t(size)
  current_time = (Time.now - start_time).round(5)
  puts "Allocating memory for output stream on server:\t#{current_time}s"

  # Write to LMem
  start_time = Time.now
  actions_write_lmem = LMemLoopback_writeLMem_actions_t_struct.new
  actions_write_lmem.param_address = 0
  actions_write_lmem.param_nbytes = size_bytes
  actions_write_lmem.instream_cpu_to_lmem = address_in_a
  address_actions_write_lmem =
    client.send_LMemLoopback_writeLMem_actions_t(actions_write_lmem)
  client.LMemLoopback_writeLMem_run(max_engine, address_actions_write_lmem)

  actions_write_lmem.param_address = size_bytes
  actions_write_lmem.param_nbytes = size_bytes
  actions_write_lmem.instream_cpu_to_lmem = address_in_b
  address_actions_write_lmem =
    client.send_LMemLoopback_writeLMem_actions_t(actions_write_lmem)
  client.LMemLoopback_writeLMem_run(max_engine, address_actions_write_lmem)
  current_time = (Time.now - start_time).round(5)
  puts "Writing to LMem:\t\t\t\t#{current_time}s"

  # Action default
  start_time = Time.now
  actions = LMemLoopback_actions_t_struct.new
  actions.param_N = size

  address_actions = client.send_LMemLoopback_actions_t(actions)
  client.LMemLoopback_run(max_engine, address_actions)
  current_time = (Time.now - start_time).round(5)
  puts "LMemLoopback time:\t\t\t\t#{current_time}s"

  # Read from LMem
  start_time = Time.now
  actions_read_lmem = LMemLoopback_readLMem_actions_t_struct.new

  actions_read_lmem.param_address = 2 * size_bytes
  actions_read_lmem.param_nbytes = size_bytes
  actions_read_lmem.outstream_lmem_to_cpu = address_out_data

  address_actions_read_lmem =
    client.send_LMemLoopback_readLMem_actions_t(actions_read_lmem)
  client.LMemLoopback_readLMem_run(max_engine, address_actions_read_lmem)
  current_time = (Time.now - start_time).round(5)
  puts "Reading from LMem:\t\t\t\t#{current_time}s"

  # Unload DFE
  start_time = Time.now
  client.max_unload(max_engine)
  current_time = (Time.now - start_time).round(5)
  puts "Unloading DFE:\t\t\t\t\t#{current_time}s"

  # Get output stream from server
  start_time = Time.now
  out_data = client.receive_data_int32_t(address_out_data, size)
  current_time = (Time.now - start_time).round(5)
  puts "Getting output stream:\t(size = #{size * 32} bit)\t#{current_time}s"

  # Free allocated memory for streams on server
  start_time = Time.now
  client.free(address_in_a)
  client.free(address_in_b)
  client.free(address_out_data)
  client.free(address_actions_write_lmem)
  client.free(address_actions)
  client.free(address_actions_read_lmem)
  current_time = (Time.now - start_time).round(5)
  puts "Freeing allocated memory for streams on server:\t#{current_time}s"

  # Free allocated maxfile data
  start_time = Time.now
  client.LMemLoopback_free
  current_time = (Time.now - start_time).round(5)
  puts "Freeing allocated maxfile data:\t\t\t#{current_time}s"

  # Close!
  start_time = Time.now
  transport.close
  current_time = (Time.now - start_time).round(5)
  puts "Closing connection:\t\t\t\t#{current_time}s"

  current_time = (Time.now - start_dfe_time).round(5)
  puts "DFE LMemLoopback total time:\t\t\t#{current_time}s"

  # CPU Output
  start_time = Time.now
  expected = Array.new(size)
  lmem_loopback_cpu(in_b, in_a, size, expected)
  current_time = (Time.now - start_time).round(5)
  puts "CPU LMemLoopback total time:\t\t\t#{current_time}s"

  # Checking results
  start_time = Time.now
  status = check(out_data, expected, size)
  current_time = (Time.now - start_time).round(5)
  puts "Checking results:\t\t\t\t#{current_time}s"

  if status == 0
    puts 'Test passed!'
  else
    puts "Test failed #{status} times!"
    Kernel.exit(-1)
  end

rescue Thrift::Exception => tx
  puts 'Thrift::Exception: ', tx.message, "\n"
  Kernel.exit(-1)
end

File: examples/LMemLoopback/client/perl/AdvancedStatic/LMemLoopbackClient.pl
#!/usr/bin/perl -w

use strict;
use warnings;

use lib '../gen-perl';

use Thrift;
use Thrift::BinaryProtocol;
use Thrift::Socket;
use Thrift::BufferedTransport;

use com::maxeler::LMemLoopback::LMemLoopbackService;
use com::maxeler::LMemLoopback::Types;

use Data::Dumper;

use Time::HiRes qw/ time /;

sub check {
    my($dataOutDFE, $dataOutCPU, $size) = @_;
    my $status = 0;

    for(my $i = 0; $i < $size; $i++) {
        if($$dataOutDFE[$i] != $$dataOutCPU[$i]) {
            print "Output data @ ".$i." = ".$$dataOutDFE[$i]." (expected ".$$dataOutCPU[$i].")\r\n";
            $status++;
        }
    }
    return $status;    
}

sub LMemLoopbackCPU {
    my ($size, $inA, $inB) = @_;
    my @dataOut = ();

    for (my $i = 0; $i < $size; $i++) {
        $dataOut[$i] = $$inA[$i] + $$inB[$i];
    }

    return \@dataOut;
}

sub LMemLoopbackDFE {
    my ($size, $inA, $inB) = @_;
    my @outData = ();
    my $sizeBytes = $size * 4;
    eval {
	my $startTime = time;

        # Make socket
        my $socket    = new Thrift::Socket('localhost', 9090);

        # Buffering is critical. Raw sockets are very slow
        my $transport = new Thrift::BufferedTransport($socket);

        # Wrap in a protocol
        my $protocol  = new Thrift::BinaryProtocol($transport);

        # Create a client to use the protocol encoder
	$startTime = time;
        my $client    = new com::maxeler::LMemLoopback::LMemLoopbackServiceClient($protocol);
	print "Creating a client:\t\t\t\t", (time - $startTime), "s\n";

        # Connect!
	$startTime = time;
        $transport->open();
	print "Opening connection:\t\t\t\t", (time - $startTime), "s\n";

        # Initialize maxfile
	$startTime = time;
        my $max_file = $client->LMemLoopback_init();
	print "Initializing maxfile:\t\t\t\t", (time - $startTime), "s\n";

        # Load DFE
	$startTime = time;
        my $engine = $client->max_load($max_file, '*');
	print "Loading DFE:\t\t\t\t\t", (time - $startTime), "s\n";

        # Allocate and send input streams to server
	$startTime = time;
        my $address_inA = $client->malloc_int32_t($size);
        $client->send_data_int32_t($address_inA, $inA);

        my $address_inB = $client->malloc_int32_t($size);
        $client->send_data_int32_t($address_inB, $inB);
	print "Sending input data:\t\t\t\t", (time - $startTime), "s\n";

        # Allocate memory for output stream on server
	$startTime = time;
        my $address_outData = $client->malloc_int32_t($size);
        print "Allocating memory for output stream on server:\t", (time - $startTime), "s\n";

        # Write to LMem
	$startTime = time;
        my $actions_lmem_write = new com::maxeler::LMemLoopback::LMemLoopback_writeLMem_actions_t_struct();

        $actions_lmem_write->{param_address} = 0;
        $actions_lmem_write->{param_nbytes} = $sizeBytes;
        $actions_lmem_write->{instream_cpu_to_lmem} = $address_inA;

        my $address_actions_lmem_write = $client->send_LMemLoopback_writeLMem_actions_t($actions_lmem_write);
        $client->LMemLoopback_writeLMem_run($engine, $address_actions_lmem_write);

        $actions_lmem_write->{param_address} = $sizeBytes;
        $actions_lmem_write->{param_nbytes} = $sizeBytes;
        $actions_lmem_write->{instream_cpu_to_lmem} = $address_inB;

        $address_actions_lmem_write = $client->send_LMemLoopback_writeLMem_actions_t($actions_lmem_write);
        $client->LMemLoopback_writeLMem_run($engine, $address_actions_lmem_write);
        $client->free($address_actions_lmem_write);
	print "Writing to LMem:\t\t\t\t", (time - $startTime), "s\n";

        # Action default
	$startTime = time;
        my $actions = new com::maxeler::LMemLoopback::LMemLoopback_actions_t_struct();

        $actions->{param_N} = $size;

        my $address_actions = $client->send_LMemLoopback_actions_t($actions);
        $client->LMemLoopback_run($engine, $address_actions);
        $client->free($address_actions);
	print "LMemLoopback time:\t\t\t\t", (time - $startTime), "s\n";

        # Read from LMem
	$startTime = time;
        my $actions_lmem_read = new com::maxeler::LMemLoopback::LMemLoopback_readLMem_actions_t_struct();

        $actions_lmem_read->{param_address} = 2 * $sizeBytes;
        $actions_lmem_read->{param_nbytes} = $sizeBytes;
        $actions_lmem_read->{outstream_lmem_to_cpu} = $address_outData;

        my $address_actions_lmem_read = $client->send_LMemLoopback_readLMem_actions_t($actions_lmem_read);
        $client->LMemLoopback_readLMem_run($engine, $address_actions_lmem_read);
        $client->free($address_actions_lmem_read);
	print "Reading from LMem:\t\t\t\t", (time - $startTime), "s\n";

        # Unload DFE
	$startTime = time;
        $client->max_unload($engine);
	print "Unoading DFE:\t\t\t\t\t", (time - $startTime), "s\n";

        # Get output stream from server
	$startTime = time;
        my $outData_ref = $client->receive_data_int32_t($address_outData, $size);
        @outData = @$outData_ref;
	print "Getting output stream:\t(size = ", ($size * 32), " bit)\t", (time - $startTime), "s\n";

        # Free allocated memory for streams on server
	$startTime = time;
        $client->free($address_inA);
        $client->free($address_inB);
        $client->free($address_outData);
	print "Freeing allocated memory for streams on server:\t", (time - $startTime), "s\n";

        # Free allocated maxfile data
	$startTime = time;
        $client->LMemLoopback_free();
	print "Freeing allocated maxfile data:\t\t\t", (time - $startTime), "s\n";

        # Close!
	$startTime = time;
        $transport->close();
	print "Closing connection:\t\t\t\t", (time - $startTime), "s\n";

    }; if($@){
        warn(Dumper($@));
    }

    return \@outData;
}
my $startTime = time;
my $size = 384;
my @inA = ();
my @inB = ();
my @dataOutCPU = ();
my @dataOutDFE = ();

for(my $i = 0; $i < $size; $i++) {
    $inA[$i] = $i;
    $inB[$i] = $size - $i;
}
print "Generating input data:\t\t\t\t", (time - $startTime), "s\n";

# DFE Output
$startTime = time;
my $dataOutDFE_ref = LMemLoopbackDFE($size, \@inA, \@inB);
print "DFE LMemLoopback total time:\t\t\t", (time - $startTime), "s\n";

# CPU Output
$startTime = time;
my $dataOutCPU_ref = LMemLoopbackCPU($size, \@inA, \@inB);
print "CPU LMemLoopback total time:\t\t\t", (time - $startTime), "s\n";

@dataOutCPU = @$dataOutCPU_ref;
@dataOutDFE = @$dataOutDFE_ref;

# Checking results
$startTime = time;
my $status = check(\@dataOutDFE, \@dataOutCPU, $size);
print "Checking results:\t\t\t\t", (time - $startTime), "s\n";

if ($status == 0) {
    print "Test successful!\n";
} else {
        print "Test failed ", $status, " times!\n";
        exit -1;
}
 
